### ISSUE TITLE
Treasury – Staking Delegation

### ISSUE THREAD

**Issue by ori-near** [NEAR-DevHub/neardevhub-treasury-dashboard#104]
# Background
We have already built an MVP for our Treasury Dashboard with basic functions (Dashboard, Payments, Settings). This ticket focuses on adding a "Stake Delegation" feature specifically for managing the staking of liquid tokens available within the DAO. A separate feature for handling locked tokens will be introduced in a later ticket. 

# User Stories
* As a DAO member with "create" permission, I want to be able to create staking and unstaking requests so that I can manage the DAO's token delegation and earn rewards.
 * As a DAO member with "vote" permission, I want to be able to review and approve/reject staking and unstaking requests so that I can participate in the DAO's treasury governance process.
 * As a general user, I want to be able to view the history of staking and unstaking requests so that I can track the DAO's delegation activities.

# Acceptance Criteria

## Stake Delegation Page

- [x] Add a "Stake Delegation" link to the Treasury Dashboard navigation.
- [x] Clicking the link displays a page with two tabs: Pending Requests and History. The number of requests in pending history is shown next to the titles.
- [x] Users with create permission can see and click on the Create Request button dropdown, which has two options: Staking, Unstaking.
- [x] There is a filter tab that allows user to enable/disable and rearrange columns within the table below
- [x] There is pagination at the bottom to handle a large number of requests (e.g. greater than 20).

### Pending Requests Tabs

- [x] Displays a list of all pending staking and unstaking requests, including the following columns:
  - [x] ID: The ID of the pending request (default sort, ascending order from 01)
  - [x] Created Date: The date and time the request was made (in UTC)
  - [x] Amount: In NEAR and USD equivalence
  - [x] Validator: Include chosen validator URL, fee, and status
  - [x] Created By: The username of the admin who created the request
  - [x] Notes: Any notes entered by request creator
  - [x] Votes: Number of votes required for approval and current number of completed votes
  - [x] Approvers: The usernames of the users who approved or rejected and their vote indication
  - [x] Actions: Approve, Reject – Only visible to users with vote permissions
  - [x] Details – Shows transaction details
  - [ ] Pencil edit (not shown in diagram) – this button allows users with create permission to edit the request before it had any votes. Clicking on this will show the create staking or unstaking dialogue with the pre-populated information, along with a delete button option.

### Create Staking Request
- [ ] User with create permission who click on create staking request will see the dialogue with the following elements:
  - [x] Display treasury's available balance and total staked amount.
  - [x] Allow users to select a validator from a dynamically fetched list (need to clarify where this is coming from), showing the validator's account, fee, and status.
  - [x] Amount to stake input, with a warning displayed if user tries to exceed the available balance.
  - [x] Notes text area for optional comments.
  - [ ] Delete button (show this to users with create permissions only when editing an existing staking request, not during the initial creation).
  - [x] Cancel button – will clear all fields and close the dialogue.
  - [x] Submit button – triggers a confirmation screen and will then make the request appear in the "Pending Requests" table.

### Create Unstaking Request

- [x] User with create permission who click on create unsticking request will see the dialogue with the following elements:
  - [x] Display treasury's available balance and total staked amount.
  - [ ] Only show a list of selected validators with active staking for the treasury, including the validator's account, fee, status, and staked amount (NEAR and USD equivalence).
  - [x] Amount to unstake, with a warning displayed if user tries to exceed the staking balance.
  - [x] Notes text area for optional comments.
  - [ ] Delete button (show this to users with create permissions only when editing an existing staking request, not during the initial creation).
  - [x] Cancel button – will clear all fields and close the dialogue.
  - [x] Submit button – triggers a confirmation screen and will then make the request appear in the "Pending Requests" table.

### History Tab

- [x] Display a list of all past staking and unstaking requests, with their status (Approved, Expired, Rejected).
- [ ] Include the same columns as the "Pending Requests" tabs, with the addition of a "Transaction Date" and "Transaction Link" columns.

# Request Execution
- [x] Requests are executed (tokens are staked or unstated) only after the required number of approval is met, as defined in the voting policy settings
- [x] Once executed, the request moves to the "History" tab"

# Permission Considerations
- [x] All users can view the "Stake Delegation" page and both tabs.
- [x] Users with "create" permission can create staking and unstaking requests.
- [ ] Users with "create" permission can delete requests before they were approved. These requests will not show up in history.
- [x] Users with "vote" permission can approve or reject pending requests.

# Attachments

Ask @FREZZZBE for the latest mockups.

# Pending Requests Tab
![Image](https://github.com/user-attachments/assets/b525f434-3001-43bb-93af-3e43156b1857)

## Create Staking Request
![Image](https://github.com/user-attachments/assets/38a27c0d-23c2-4309-9eab-9e0dd2a587e8)

## Create Unstaking Request
![Image](https://github.com/user-attachments/assets/2f8d0f51-4b45-4725-bd8e-353992503098)

## Create Unstaking Request – No Validators
![Image](https://github.com/user-attachments/assets/d882d11c-09da-4c9b-beca-404a890763d9)

# History Tab
![Image](https://github.com/user-attachments/assets/3c87b20c-0d81-481a-bada-97d369964b23)

**Comment by Megha-Dev-19** [comment]
hey @ori-near, few things:
> ID: The ID of the pending request (default sort, ascending order from 01)

- we show the ID in descending order, so the latest request is at the top

> Amount: In NEAR and USD equivalence
- I remember we put the USD equivalence part on hold (for payments as well), since when the request was created the NEAR can be of different price, and when the user view it, it will be different, we decided to add it after the indexer (I will add this point to the indexer ticket).

> Pencil edit (not shown in diagram) – this button allows users with create permission to edit the request before it had any votes. Clicking on this will show the create staking or unstaking dialogue with the pre-populated information, along with a delete button option.
> Delete button (show this to users with create permissions only when editing an existing staking request, not during the initial creation).
> Users with "create" permission can delete requests before they were approved. These requests will not show up in history
- can I remove this, since this a part of [separate ticket](https://github.com/orgs/NEAR-DevHub/projects/4/views/21?pane=issue&itemId=85677919)?

> Include the same columns as the "Pending Requests" tabs, with the addition of a "Transaction Date" and "Transaction Link" columns.
- can I remove this, since this a part of [ticket](https://github.com/NEAR-DevHub/neardevhub-treasury-dashboard/issues/107)
